packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +464,127 @@ export default {
     // host-by-host install
     throw new Error('non pool-wide install not implemented')
   },
+
+  async rollingPoolUpdate() {
+    let hosts = filter(this.objects.all, { $type: 'host' })
+    await Promise.all(hosts.map(host => this.call('host.assert_can_evacuate', host.$ref)))
+
+    log.debug('Install patches')
+    await this.installPatches()
+
+    // Remember on which hosts and SRs the running VMs are
+    const vmsByHost = mapValues(
+      groupBy(
+        filter(this.objects.all, {
+          $type: 'VM',
+          power_state: 'Running',
+          is_control_domain: false,
+        }),
+        vm => vm.$resident_on?.$id
+      ),
+      vms =>
+        vms.map(vm => {
+          const mapVifsNetworks = {}
+          vm.$VIFs.forEach(vif => {
+            mapVifsNetworks[vif.$id] = vif.$network.$id
+          })
+
+          const mapVdisSrs = {}
+          flatMap(vm.$snapshots, '$VBDs')
+            .concat(vm.$VBDs)
+            .forEach(vbd => {
+              if (vbd.type === 'Disk') {
+                const vdi = vbd.$VDI
+                mapVdisSrs[vdi.$id] = vdi.$SR.$id
+              }
+            })
+
+          return {
+            id: vm.$id,
+            ref: vm.$ref,
+            mapVifsNetworks,
+            mapVdisSrs,
+          }
+        })
+    )
+
+    if (vmsByHost.undefined !== undefined) {
+      throw new Error('Could not find host of all running VMs')
+    }
+
+    // Cache $fields to make sure we can access them later
+    hosts = hosts.map(host => ({
+      ...host,
+      id: host.$id,
+      ref: host.$ref,
+      metricsId: host.$metrics.$id,
+      metricsRef: host.$metrics.$ref,
+    }))
+
+    // Put master in first position to restart it first
+    const indexOfMaster = hosts.findIndex(host => host.ref === this.pool.master)
+    if (indexOfMaster > 0) {
+      ;[hosts[0], hosts[indexOfMaster]] = [hosts[indexOfMaster], hosts[0]]
+    }
+
+    // Restart all the hosts one by one
+    for (const host of hosts) {
+      const hostId = host.id
+      const metricsId = host.metricsId
+
+      await this.barrier(host.metricsRef)
+      await this._waitObjectState(metricsId, metrics => metrics.live)
+      const rebootTime = parseDateTime(await this.call('host.get_servertime', host.ref))
+
+      log.debug(`Evacuate and restart host ${hostId}`)
+      await this.rebootHost(hostId)
+
+      log.debug(`Wait for host ${hostId} to be up`)
+      await Promise.race([
```

Comment:
You can use [`promise-toolbox/timeout`](https://github.com/JsCommunity/promise-toolbox#promisetimeoutms-cb-or-rejectionvalue).